### PR TITLE
Kernel runtime error checking

### PR DESCRIPTION
**Pull request by nashif**
Introduce runtime error checking in the kernel where it make sense and remove duplication of asserts and oopses that were used to catch such errors. Propagate errors where needed. 

Ideally we want to use this to check for parameters and permissions and return to the caller to allow recovery, which was not possible in the current approach: when asserts are enabled (not default) we would just crash on every error, even if recoverable or we will get undefined behavior if asserts were not enabled.

Situations which should never happen and disallowed by design, such as running something in an ISR when it should not, remain as asserts where applicable.

-----------------

(Edit by @aescolar) Discussed in #17105

### CODE REVIEW

File: include/kernel.h [hunk 1]
```diff
@@ -3030,10 +3030,12 @@ __syscall int k_mutex_lock(struct k_mutex *mutex, s32_t timeout);
  *
  * @param mutex Address of the mutex.
  *
- * @return N/A
+ * @retval 0 Mutex unlocked.
+ * @retval -EPERM The current thread does not own the mutex.
+ * @retval -EINVAL Undefined behavior.
```
Comment (ceolin) [inline include/kernel.h]: EINVAL is for invalid parameter not undefined behavior

Comment (ceolin) [reply include/kernel.h]: btw, return EINVAL seems fine, can't we just say that mutex is not locked

File: kernel/poll.c [hunk 2]
```diff
@@ -35,13 +35,17 @@
...
+int k_poll_event_init(struct k_poll_event *event, u32_t type,
 		       int mode, void *obj)
 {
-	__ASSERT(mode == K_POLL_MODE_NOTIFY_ONLY,
-		 "only NOTIFY_ONLY mode is supported\n");
-	__ASSERT(type < (BIT(_POLL_NUM_TYPES)), "invalid type\n");
-	__ASSERT(obj != NULL, "must provide an object\n");
+	/* only NOTIFY_ONLY mode is supported */
+	if (mode != K_POLL_MODE_NOTIFY_ONLY) {
+		return -ENOTSUP;
```
Comment (pfl) [inline kernel/poll.c]: -ENOTSUP is not documented in the .h file?

File: kernel/queue.c [hunk 3]
```diff
@@ -255,11 +258,18 @@ void k_queue_append_list(struct k_queue *queue, void *head, void *tail)
...
-void k_queue_merge_slist(struct k_queue *queue, sys_slist_t *list)
+int k_queue_merge_slist(struct k_queue *queue, sys_slist_t *list)
 {
-	__ASSERT(!sys_slist_is_empty(list), "list must not be empty");
+	int ret;
+
+	/* list must not be empty */
+	if (sys_slist_is_empty(list)) {
+		return -EINVAL;
+	}
```
Comment (pfl) [inline kernel/queue.c]: Nitipick: if the list is empty, the function could just return here?

**Review by pfl — APPROVED** [review]

File: include/kernel.h (hunk 1 above)
Comment (nashif) [reply include/kernel.h]: I am going to follow posix here and return EPERM in both cases

File: kernel/mem_slab.c [hunk 4]
```diff
@@ -28,15 +28,16 @@ struct k_mem_slab *_trace_list_k_mem_slab;
  *
  * @return N/A
  */
-static void create_free_list(struct k_mem_slab *slab)
+static int create_free_list(struct k_mem_slab *slab)
```
Comment (andrewboie) [inline kernel/mem_slab.c]: not all callers of this function are checking the return value.

File: kernel/msg_q.c [hunk 5]
```diff
@@ -94,14 +94,17 @@ Z_SYSCALL_HANDLER(k_msgq_alloc_init, q, msg_size, max_msgs)
 }
 #endif
 
-void k_msgq_cleanup(struct k_msgq *q)
+int k_msgq_cleanup(struct k_msgq *q)
 {
-	__ASSERT_NO_MSG(z_waitq_head(&q->wait_q) == NULL);
+	if (z_waitq_head(&q->wait_q) != NULL) {
+		return -EAGAIN;
```
Comment (andrewboie) [inline kernel/msg_q.c]: maybe -EBUSY would be more fitting?

File: kernel/mutex.c [hunk 6]
```diff
@@ -200,18 +182,32 @@ Z_SYSCALL_HANDLER(k_mutex_lock, mutex, timeout)
...
+	/*
+	 * If a thread attempts to unlock a mutex that it has not locked or a
+	 * mutex which is unlocked, undefined behavior results.
+	 *
+	 * Similar to posix, if a thread attempts to unlock a mutex which is
+	 * unlocked, k_mutex_unlock() shall behave as "Unlock When Not Owner"
+	 * below.
+	 */
+	if (mutex->lock_count == 0U) {
+		return -EPERM;
```
Comment (andrewboie) [inline kernel/mutex.c]: Please use the same error codes as sys_mutex(), for consistency.
This should be -EINVAL

**Review by tyrantlizard77 — CHANGES_REQUESTED** [review]
You can get rid of some of the checks in sys_mutex code since the base implementation is now doing it.

File: kernel/mutex.c [hunk 7]
```diff
@@ -203,12 +203,24 @@ static inline int z_vrfy_k_mutex_lock(struct k_mutex *mutex, s32_t timeout)
...
 #endif
 
-void z_impl_k_mutex_unlock(struct k_mutex *mutex)
+int z_impl_k_mutex_unlock(struct k_mutex *mutex)
 {
 	struct k_thread *new_owner;
 
-	__ASSERT(mutex->lock_count > 0U, "");
-	__ASSERT(mutex->owner == _current, "");
+	/*
```
Comment (dcpleung) [inline kernel/mutex.c]: I think permission check should be first so it won't leak information about a non-owned mutex.

File: kernel/pipes.c [hunk 8]
```diff
@@ -173,16 +173,19 @@ static inline int z_vrfy_k_pipe_alloc_init(struct k_pipe *pipe, size_t size)
 #include <syscalls/k_pipe_alloc_init_mrsh.c>
 #endif
 
-void k_pipe_cleanup(struct k_pipe *pipe)
+int k_pipe_cleanup(struct k_pipe *pipe)
 {
-	__ASSERT_NO_MSG(!z_waitq_head(&pipe->wait_q.readers));
-	__ASSERT_NO_MSG(!z_waitq_head(&pipe->wait_q.writers));
+	if (z_waitq_head(&pipe->wait_q.readers) ||
+			z_waitq_head(&pipe->wait_q.writers)) {
```
Comment (ceolin) [inline kernel/pipes.c]: can you make it boolean operations ? e.g `z_waitq_head(&pipe->wait_q.readers) != NULL`

File: include/kernel.h [hunk 9]
```diff
@@ -1543,9 +1543,10 @@ typedef void (*k_timer_stop_t)(struct k_timer *timer);
  * @param expiry_fn Function to invoke each time the timer expires.
  * @param stop_fn   Function to invoke if the timer is stopped while running.
  *
- * @return N/A
+ * @retval 0 on success
```
Comment (ceolin) [inline include/kernel.h]: why change it if this function always return 0 ? There is nothing to check here.

File: kernel/mutex.c [hunk 10]
```diff
@@ -130,10 +124,6 @@ int z_impl_k_mutex_lock(struct k_mutex *mutex, s32_t timeout)
 		mutex->lock_count++;
 		mutex->owner = _current;
 
-		K_DEBUG("%p took mutex %p, count: %d, orig prio: %d\n",
-			_current, mutex, mutex->lock_count,
-			mutex->owner_orig_prio);
-
```
Comment (ceolin) [inline kernel/mutex.c]: +1 for removing K_DEBUG, but wouldn't be good keep this info ? we could call printk directly under `ifdef CONFIG_DEBUG` guard.

File: include/kernel.h [hunk 11]
```diff
@@ -3448,11 +3448,19 @@ void k_msgq_init(struct k_msgq *q, char *buffer, size_t msg_size,
...
 
-void k_msgq_cleanup(struct k_msgq *q);
+/**
+ * @brief Release allocated buffer for a queue
+ *
+ * @param msgq message queue to cleanup
+ *
+ * @retval 0 on success
+ * @retval -EAGAIN Queue not empty
+ */
```
Comment (ceolin) [inline include/kernel.h]: s/EAGAIN/EBUSY/

Comment (ceolin) [reply include/kernel.h]: or change the return in the implementation :) but EBUSY seems better

Comment (ceolin) [reply include/kernel.h]: though `k_stack_cleanup` is returning EAGAIN. Just need to be consistent :)

**Review by aescolar — DISMISSED** [review]
@nashif I can certainly understand the motivation for going in this direction. But doing the change like this leaves behind both:
* smaller embedded systems which will be too constrained to do all error checking in the app
* current systems which rely on asserts at runtime (Which would lose that error checking until their apps are rewritten)

On https://github.com/zephyrproject-rtos/zephyr/issues/17105#issuecomment-508694903 I elaborated a bit more on the problem and on the kind of solution I really think would be best.
But if that is too cumbersome to implement, I would at least allow to kconfig if there is asserts or runtime checks, instead of just leaving only one option.

**Comment by nashif** [comment]
> @nashif I can certainly understand the motivation for going in this direction. But doing the change like this leaves behind both:
> 
>     * smaller embedded systems which will be too constrained to do all error checking in the app
> 
>     * current systems which rely on asserts at runtime (Which would lose that error checking until their apps are rewritten)

@aescolar there is no intention to go and remove every assert we have and replace it with runtime error-checking. However, we do need to go through the usages of asserts and create some guidelines, because asserts in general are being misused and we are left in some cases with no runtime checking where it really matters, because someone decided it is easier to just use an ASSERT or because it has always been done this way, etc.

Regarding footprint in general we need to be talking in numbers here and not talk in abstracts ("smaller embedded systems...."), i.e. set targets, have budgets and put everyone and every subsystem under the same rules and guidelines and scrutiny. I am sure there is always more to be saved in terms of footprint above the kernel.

The kernel is not the only place where footprint should be saved, it is however the most importat area where runtime error checking is needed due to the fact that it can be the source for many issues showing up all the way up the stack if not handled correctly, both from quality, safety and security perspective.

**Comment by ceolin** [comment]
> > @nashif I can certainly understand the motivation for going in this direction. But doing the change like this leaves behind both:
> > ```
> > * smaller embedded systems which will be too constrained to do all error checking in the app
> > 
> > * current systems which rely on asserts at runtime (Which would lose that error checking until their apps are rewritten)
> > ```
> 
> @aescolar there is no intention to go and remove every assert we have and replace it with runtime error-checking. However, we do need to go through the usages of asserts and create some guidelines, because asserts in general are being misused and we are left in some cases with no runtime checking where it really matters, because someone decided it is easier to just use an ASSERT or because it has always been done this way, etc.
> 
> Regarding footprint in general we need to be talking in numbers here and not talk in abstracts ("smaller embedded systems...."), i.e. set targets, have budgets and put everyone and every subsystem under the same rules and guidelines and scrutiny. I am sure there is always more to be saved in terms of footprint above the kernel.
> 
> The kernel is not the only place where footprint should be saved, it is however the most importat area where runtime error checking is needed due to the fact that it can be the source for many issues showing up all the way up the stack if not handled correctly, both from quality, safety and security perspective.

agreed ! Adding two more cents, don't have this kind of error checking within the kernel makes impossible do it properly on up layers. For instance, syscalls can not return proper errors if we don't have it done in inner functions.

**Comment by aescolar** [comment]
@nashif : Ok, so given that this is WIP, and I cannot read your mind to know what is it that you plan to do,
instead of me trying to guess the extent of what you want to do, and probably talking past each other, can you describe what is it you are proposing?

My concerns and arguments I described on #17105 (comment). The main argument is: _depending on what you are trying to build, your needs can be very different_.

**Comment by andrewboie** [comment]
> smaller embedded systems which will be too constrained to do all error checking in the app

Can we be data-driven here, and show some numbers on the footprint increase associated with this, before we sound the alarm?

**Comment by aescolar** [comment]
@andrewboie @nashif 
Guys, it seems to me you are fixated thinking this is an "either or".
Doesn't doing something like this below cover everybody's needs?
(Please ignore naming)
```diff
diff --git a/kernel/sem.c b/kernel/sem.c
index f1fd430e64..c324bc315b 100644
--- a/kernel/sem.c
+++ b/kernel/sem.c
@@ -59,13 +59,27 @@ SYS_INIT(init_sem_module, PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_OBJECTS);
 
 #endif /* CONFIG_OBJECT_TRACING */
 
+
+#if defined(ASSERT_ON_ERRORS_IN_KERNEL)
+       #define CHECK(expr) \
+                __ASSERT_NO_MSG(!(expr));   \
+                if (0)
+#elif defined(RETURN_ERRORS_IN_KERNEL)
+       #define CHECK(expr) \
+                if (expr)
+#elif defined(NO_KERNEL_RUNTIME_CHECKS)
+       #define CHECK(...) \
+                if (0)
+#endif
+
+
 int z_impl_k_sem_init(struct k_sem *sem, unsigned int initial_count,
                      unsigned int limit)
 {
        /*
         * Limit cannot be zero and count cannot be greater than limit
         */
-       if (limit == 0U || initial_count > limit) {
+       CHECK(limit == 0U || initial_count > limit) {
                return -EINVAL;
        }
```

**Comment by andrewboie** [comment]
> Doesn't doing something like this below cover everybody's needs? (Please ignore naming)

Being specific really helps. This approach looks good to me. We'll need to ensure that if USERSPACE=y then the RETURN_ERRORS_IN_KERNEL option is selected, since the syscall verifiers will rely on them.

**Review by tyrantlizard77 — CHANGES_REQUESTED** [review]
At this time I concur with @aescolar now that he has proposed an alternative to doing this which gives greater control on how much footprint to use for error handling.

**Comment by nashif** [comment]
using a CHECK macro now.

File: Kconfig.zephyr [hunk 12]
```diff
@@ -248,6 +248,30 @@ config COMPILER_OPT
 
 endmenu
 
+
+
```
Comment (galak) [inline Kconfig.zephyr]: nit, extra whitespace/newline.

File: include/kernel.h [hunk 13]
```diff
@@ -3586,11 +3598,19 @@ void k_msgq_init(struct k_msgq *q, char *buffer, size_t msg_size,
  *	an integer overflow.
  * @req K-MSGQ-002
  */
-__syscall int k_msgq_alloc_init(struct k_msgq *q, size_t msg_size,
+__syscall int k_msgq_alloc_init(struct k_msgq *msgq, size_t msg_size,
```
Comment (galak) [inline include/kernel.h]: can you pull the arg renaming into a different PR/commit?

File: Kconfig.zephyr [hunk 14]
```diff
@@ -248,6 +248,30 @@ config COMPILER_OPT
...
+	help
+	  Assert on errors covered with the CHECK macro.
+
+config NO_RUNTIME_CHECKS
+	bool "No runtime checks"
+	help
+	  Do not do any runtime checks or asserts.
+
+config RUNTIME_ERROR_CHECKS
+	bool "No runtime checks"
```
Comment (galak) [inline Kconfig.zephyr]: the 'no runtimes checks' seems wrong here.

File: include/kernel.h [hunk 15]
```diff
@@ -3610,7 +3630,7 @@ void k_msgq_cleanup(struct k_msgq *q);
  * @retval -EAGAIN Waiting period timed out.
  * @req K-MSGQ-002
  */
-__syscall int k_msgq_put(struct k_msgq *q, void *data, s32_t timeout);
+__syscall int k_msgq_put(struct k_msgq *msgq, void *data, s32_t timeout);
```
Comment (galak) [inline include/kernel.h]: similar comment about arg renaming, just makes this a bit more difficult to review.

**Review by galak — DISMISSED** [review]
Minor issue in the Kconfig that needs cleaning up.

File: Kconfig.zephyr [hunk 16]
```diff
@@ -248,6 +248,27 @@ config COMPILER_OPT
...
+
+config NO_RUNTIME_CHECKS
+	bool "No runtime error checks"
+	help
+	  Do not do any runtime checks or asserts.
+
+config RUNTIME_ERROR_CHECKS
+	bool "Enable runtime error checks"
+	help
+	  Do runtime checks, this is the default and the only option used
```
Comment (aescolar) [inline Kconfig.zephyr]: Nit: Maybe better rephrase "and the only option used" => "and the typical option used"
I guess we will need sometimes to disable input checks for some reasons (e.g. sample almost too big for microbit)

File: Kconfig.zephyr [hunk 17]
```diff
@@ -248,6 +248,27 @@ config COMPILER_OPT
...
+
+config ASSERT_ON_ERRORS
+	bool "Assert on all errors"
+	help
+	  Assert on errors covered with the CHECK macro.
+
+config NO_RUNTIME_CHECKS
+	bool "No runtime error checks"
+	help
+	  Do not do any runtime checks or asserts.
```
Comment (aescolar) [inline Kconfig.zephyr]: Nit: How it is phrased now, it seems to imply this option disables all asserts and error checking everywhere.

Would it be better to mention in the prompt above "Error checking behavior for CHECK macro" (and remove "covered with the CHECK macro" from the first option).
OR
repeat "covered with the CHECK macro" in the 2 others?

**Review by aescolar — COMMENTED** [review]
Thanks Anas, this should cover both needs nicely.
Just two minor comments

File: Kconfig.zephyr (hunk 16 above)
Comment (nashif) [reply Kconfig.zephyr]: This is going to be the only option enabled during testing, there is no way around it. Enable any of the other options and things will collapse now that we are going to do negative testing, the whole purpose of this action.

File: Kconfig.zephyr (hunk 17 above)
Comment (nashif) [reply Kconfig.zephyr]: sure.

File: include/sys/check.h [hunk 18]
```diff
@@ -0,0 +1,26 @@
...
+#define ZEPHYR_INCLUDE_SYS_CHECK_H_
+
+#include <sys/__assert.h>
+
+#if defined(CONFIG_ASSERT_ON_ERRORS)
+#define CHECK(expr) \
+	__ASSERT_NO_MSG(!(expr));   \
+	if (0)
+#elif defined(CONFIG_NO_RUNTIME_CHECKS)
+#define CHECK(...) \
```
Comment (andrewboie) [inline include/sys/check.h]: Are there any cases where completely dropping the arguments could generate a compiler warning about "set but not used" or something like that?

**Comment by andrewboie** [comment]
Any particular reason why the assertions in kernel/mem_domain.c were not updated with checks?

One final thought: what do we do about k_thread_create()?

**Comment by nashif** [comment]
> One final thought: what do we do about k_thread_create()?

what about it? there is one assert in there that should be left as such (creating thread from ISR). Are you referring to the load of Z_OOPSes in z_vrfy_k_thread_create?

**Comment by nashif** [comment]
> Any particular reason why the assertions in kernel/mem_domain.c were not updated with checks?

No reason, first I wanted us to agree this approach is the one to follow before moving too fast, there is also poll.c which also needs conversion.

**Comment by andrewboie** [comment]
> what about it? there is one assert in there that should be left as such (creating thread from ISR). Are you referring to the load of Z_OOPSes in z_vrfy_k_thread_create?

Yeah -- I'd rather propagate EINVAL somehow.
But since it returns a pointer...maybe it's enough to return NULL if a check fails.

**Review by tyrantlizard77 — APPROVED** [review]
Need to get CI passing, and please verify that choices won't inadvertently cause compiler warnings, otherwise this approach LGTM.

**Comment by nashif** [comment]
> For some weird reason this very old review is still here. It does not apply anymore. The later review superseded it

so, you are not approving?

**Review by aescolar — DISMISSED** [review]
> so, you are not approving?

I'm not the kernel maintainer, so it is for me to give a +2.
My concerns have been addressed. So this is just a +0.

File: kernel/mem_slab.c [hunk 19]
```diff
@@ -57,31 +60,45 @@ static void create_free_list(struct k_mem_slab *slab)
...
+	int rc = 0;
+
 	slab->num_blocks = num_blocks;
 	slab->block_size = block_size;
 	slab->buffer = buffer;
 	slab->num_used = 0U;
-	create_free_list(slab);
+	rc = create_free_list(slab);
+	if (rc < 0) {
+		goto out;
```
Comment (ioannisg) [inline kernel/mem_slab.c]: early return here, instead?

**Review by ioannisg — COMMENTED** [review]
some minor comments from my side

**Comment by andrewboie** [comment]
> My concerns have been addressed. So this is just a +0.

@aescolar this isn't helpful.
You obviously have strong opinions about this topic, you are allowed to give a +1

**Comment by aescolar** [comment]
> you are allowed to give a +1

@andrewboie Thanks. I understand I'm allowed to approve the PR. But the thing is that I do not support the overall direction things are going, but I will not oppose it either while the inconvenience for my company or others trying to target constrained platforms are not too great.

**Comment by ioannisg** [comment]
> > you are allowed to give a +1
> 
> @andrewboie Thanks. I understand I'm allowed to approve the PR. But the thing is that I do not support the overall direction things are going, but I will not oppose it either while the inconvenience for my company or others trying to target constrained platforms are not too great.

@aescolar @nashif as far I understood, it is possible for this patch to have no overhead, if the introduced Kconfig switches are set properly, i.e. disable both ASSERT and NO_RUNTIME error checking, am I correct?

**Comment by aescolar** [comment]
@ioannisg My concern is two fold:
* My approval of this PR may be miss-represented as support of further changes. And it is not clear to me what will happen after.
* Already @nashif is implying that testing with anything else than RUNTIME_ERROR_CHECKS will not be done.

**Comment by andrewboie** [comment]
>  But the thing is that I do not support the overall direction things are going

What? Be specific

**Comment by aescolar** [comment]
> What? Be specific

Disregard for platform constrained targets.
Over-prioritization of some assumed to be requirements of functional safety at the expense of any other target for the project.

**Comment by nashif** [comment]
> Over-prioritization of some assumed to be requirements of functional safety at the expense of any other target for the project.

This has nothing to do with functional safety. The fact that we are not able to test basic APIs because we just crash is bad for EVERYONE.

**Comment by andrewboie** [comment]
> Disregard for platform constrained targets.

@aescolar let's explore this.
How does this PR disregard platform constrained targets?
Be specific.

**Comment by aescolar** [comment]
> @aescolar let's explore this.
> How does this PR disregard platform constrained targets?

The fact that I removed my -1 would be an indication that it does not anymore. Wouldn't it?

File: Kconfig.zephyr (hunk 16 above)
Comment (aescolar) [reply Kconfig.zephyr]: If this is indeed the only option used during testing, things will probably be broken quite soon for the other options. I would make sense to do most of the kernel testing with this one. But still, I would consider it reasonable to have a couple of tests, if only tests with of samples or higher functionality, which would use the other options.
